### PR TITLE
IGNITE-22158: Sql. incorrect result of NATURAL JOIN for BIGINTEGER column on condition

### PR DESCRIPTION
**Pull request by lowka**
Rejects `NATURAL` and `USING` join conditions, if common column types do not match:
 
- Types do not match, if they represent different `ColumnType`s.
- Types that match, but have different nullability can be safely compared.
- Complex types (such as Structs, Arrays, etc. that are not fully supported by the SQL engine) never match.

https://issues.apache.org/jira/browse/IGNITE-22158

Renamed `natural_join.test` to `test_natural_join.test` and combined it with `natural_join.test_ignore`.

---

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/sql-engine/src/integrationTest/sql/join/natural/natural_join.test
Comment (lowka) [inline modules/sql-engine/src/integrationTest/sql/join/natural/natural_join.test:1]: This file was renamed to `test_natural_join.test`

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItJoinTest.java [hunk 1]
```diff
@@ -1073,4 +1077,36 @@ private static Stream<Arguments> joinTypes() {
 
         return types;
     }
+
+    @Test
+    public void testNaturalJoinTypeMismatch() {
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItJoinTest.java:1082]: I propose to mention that is temporary issue and mention a ticket here which should fix the issue.

File: modules/sql-engine/src/integrationTest/sql/join/natural/test_natural_join.test [hunk 2]
```diff
@@ -48,24 +46,28 @@ SELECT * FROM t1 NATURAL JOIN t3
...
-----
-1	2	3
-
 # common columns moved to the first place, column order: a, c, b
 query III
 SELECT * FROM t3 NATURAL JOIN t2
 ----
 1	3	2
 
+# skipif ignite3
```
Comment (ygerzhedovich) [inline modules/sql-engine/src/integrationTest/sql/join/natural/test_natural_join.test:55]: why it skipped? we have the same types for all tables

Comment (lowka) [reply modules/sql-engine/src/integrationTest/sql/join/natural/test_natural_join.test]: Thanks. Uncommented skipif instruction.

File: modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItJoinTest.java (hunk 1 above)
Comment (lowka) [reply modules/sql-engine/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItJoinTest.java:1083]: Done.

File: modules/sql-engine/src/integrationTest/sql/join/natural/test_natural_join.test [hunk 3]
```diff
@@ -48,24 +46,28 @@ SELECT * FROM t1 NATURAL JOIN t3
...
-----
-1	2	3
-
 # common columns moved to the first place, column order: a, c, b
 query III
 SELECT * FROM t3 NATURAL JOIN t2
 ----
 1	3	2
 
+skipif ignite3
```
Comment (zstan) [inline modules/sql-engine/src/integrationTest/sql/join/natural/test_natural_join.test:55]: why this skipif here ? i see that it passed in previous versions ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java [hunk 4]
```diff
@@ -182,6 +182,7 @@ public SqlNode validate(SqlNode topNode) {
 
         validateInferredDynamicParameters();
 
+
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java:185]: remove it plz

**Review by zstan — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java (hunk 4 above)
Comment (lowka) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java]: Fixed.

**Pull request merged**